password/src/Data/Password/Validate.hs

```diff
@@ -151,22 +163,32 @@ isValidPassword policy pre pass = null $ validatePassword policy pre pass
 -- []
 validatePassword :: PasswordPolicy -> CharSetPredicate -> Password -> [InvalidReason]
 validatePassword passwordPolicy@PasswordPolicy{..} charSetPredicate (Password password) =
-  catMaybes
-    [ isValidPolicy,
-      isTooShort,
-      isTooLong,
-      isUsingPolicyCharSetPredicate,
-      hasRequiredChar uppercaseChars Uppercase,
-      hasRequiredChar lowercaseChars Lowercase,
-      hasRequiredChar specialChars Special,
-      hasRequiredChar digitChars Digit
-    ]
+  -- There is no point in validating the password if either policy or predicate is invalid.
+  --
+  -- So we validate 'PasswordPolicy' and 'CharSetPredicate' first,
+  -- if they're valid, then validate 'Password'
+  firstNonEmpty
+   [ mconcat
+      [ validatePasswordPolicy passwordPolicy
+      , validateCharSetPredicate passwordPolicy charSetPredicate
+      ]
+   , catMaybes
+      [ isTooShort
+      , isTooLong
+      , isUsingValidCharacters
+      , hasRequiredChar uppercaseChars Uppercase
+      , hasRequiredChar lowercaseChars Lowercase
+      , hasRequiredChar specialChars Special
+      , hasRequiredChar digitChars Digit
+      ]
+   ]
   where
-    isValidPolicy :: Maybe InvalidReason
-    isValidPolicy =
-      if isValidPasswordPolicy passwordPolicy charSetPredicate
-        then Nothing
-        else Just $ InvalidPasswordPolicy passwordPolicy
+    -- Return first non-empty list
+    firstNonEmpty :: [[a]] -> [a]
+    firstNonEmpty [] = []
+    firstNonEmpty (x:xs)
+      | null x = firstNonEmpty xs
+      | otherwise = x
```

Comment:
I've spent a decent amount of time finding function like this within `Base`, but couldn't find it.

Reply:
[`find`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List.html#v:find) is probably what you were looking for:
```hs
find :: (a -> Bool) -> [a] -> Maybe a

firstNonEmpty = fromMaybe [] . find (not . null)
```

But seeing the code, you could also wrap both lists in a tuple and just do this:
```hs
firstNonEmpty ([], b) = b
firstNonEmpty (a, _) = a
```